### PR TITLE
v0.5.0: Mandatory Height Calibration Prompt, Compatibility with Relay Server v1.1.0, Enhanced People Panel, Various Refactors

### PR DESCRIPTION
**Pull request by Barasakar**
# Mandatory Height Calibration Prompt

Created a prototype of the height calibration prompt that asks users to calibrate their height. 

Having the height calibration done early might help addressing one of the issues (i.e., teleporting into the ground). 

Work around for #61 . 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Testing 
- [ ] Change that requires a documentation update

# How Has This Been Tested?

 Komodo runs -> prompt shows up -> calibrate height -> prompt disappears
 This is not a multi-player feature.

- [x] Manual Test
- [ ] Unit Test
- [ ] Integration / End-to-End Test

**Test Configuration**:

* _REPLACE ME: Browser vendor and version: e.g., Chrome Version 91.0.4472.164 (Official Build) (64-bit)_
* _REPLACE ME: VR Device and OS version: e.g., Oculus Quest v29_
- [ ] PC VR (Link / AirLink)
- [x] Standalone 

# Checklist:

- [ ] ~~My code follows the style guidelines of this project~~
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas that are not self-documenting
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings 
- [ ] My changes have no unnecessary logging
- [ ] I have added tests that prove my fix is effective or that my feature works, for sufficiently complex features
- [ ] ~~New and existing unit tests pass locally with my changes~~
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Sensitive info like tokens, secrets, and passwords have been removed before submitting

Modified from this article:
Phillip Johnston, “A GitHub Pull Request Template for Your Projects - Embedded Artistry,” Embedded Artistry, Aug. 04, 2017. https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ (accessed Jul. 22, 2021).

### CODE REVIEW

File: Komodo/Assets/Packages/KomodoCore/Runtime/Scripts/RuntimeSession/HeightCalibration.cs [hunk 1]
```diff
@@ -145,6 +145,8 @@ public void EndCalibration ()
 
             onFinishedCalibration.Invoke(heightToBumpPlayer);
 
+            KomodoEventManager.TriggerEvent("FinishedHeightCalibration");
```
Comment (parseccentric) [inline Komodo/Assets/Packages/KomodoCore/Runtime/Scripts/RuntimeSession/HeightCalibration.cs:148]: (future) store event names as constant strings in a central files

File: Komodo/Assets/Packages/KomodoCore/Runtime/Scripts/RuntimeSession/KomodoEventManager.cs [hunk 2]
```diff
@@ -25,24 +25,30 @@ public void Start()
...
 
+        /* This dictionary will help hold references (String) to events and events (UnityEvent) themselves */
         private Dictionary <string, UnityEvent> eventDictionary;
 
+
         private static KomodoEventManager eventManager;
 
         public static KomodoEventManager Instance
         {
+            //using the getter to find the instance
```
Comment (parseccentric) [inline Komodo/Assets/Packages/KomodoCore/Runtime/Scripts/RuntimeSession/KomodoEventManager.cs]: (future) consolidate comments

**Comment by parseccentric** [comment]
# Compatibility with Relay Server v1.1.0, Enhanced People Panel, Various Refactors

**Connect to sync namespace (breaking change)**

The client is now compatible with komodo-relay (server) v1.1.0, which requires a connection to the sync namespace.

How to test: open the Unity client in a browser and connect to a v1.1.0 server -> the people tab should report a successful connection.

Support for v1.1.0 should theoretically fix #59 

**Fix state catch-up to use messaging system (bug fix, new feature)**

The state catch-up had a bug where it would always ask the ClientSpawnManager to apply state, so networked objects (synchronized models) would not get transform, visibility, or lock states. This is now fixed.

How to test: go to Unity window > Hierarchy panel > Visible managers > NetworkManager. See Inspector panel > SocketIOEditorSimulator component, right click it, and choose ReceiveExampleStateCatchup. ->The models should all be lined up in a diagonal line, be spaced evenly, and up to four of them should be visible and locked. (Fewer will be visible and locked if some are marked as "isWholeObject = false" in ModelData. This is a limitation of the test.)

**Lock and visibility toggle infinite loop (bug fix)**

It should no longer be possible to trigger an infinite loop by rapidly pressing the lock and visibility toggle.

How to test: open two clients and rapidly press the visibility or lock button, then let go -> the button should quickly stop flashing on and off.

**Max client count (new feature)**

The maximum number of clients has arbitrarily been increased to 15.

**Improved People tab**

- The server name is displayed.
- The session number is displayed.
- The connection status is displayed.
- More connection or SocketIO room errors are displayed.
- There's now a proper grid view for client names in menu.
- The runtime app and build is now shown, rather than the configuration-time app and build. ("App" might be called "scope" or "folder" in other documentation.) Fixes #52 

Fixes #60 

**Major networking refactors**

* SocketIOJSLib
* SocketIOAdapter
* SocketIOEditorSimulator

**Refactors**

SessionStateManager has been introduced to apply multiplayer server state catch-ups to the avatars (clients) and networked objects (synchronized models).

A new class, NetworkedObjectsManager, has been introduced to manage synchronized models.

Adding and removing clients was refactored into smaller functions in an attempt to repair the out-of-body experience.

The client teardown code was moved from NetworkUpdateHandler to SocketIOAdapter for the same reason.

Minor renamings.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Testing 
- [x] Change that requires a documentation update

# How Has This Been Tested?

Please see above for the manual test instructions.

- [x] Manual Test
- [ ] Unit Test
- [x] Integration / End-to-End Test

**Test Configuration**:

* Firefox v??
* Oculus Quest v??
- [x] PC VR (Link)
- [ ] Standalone 

# Checklist:

- [ ] ~~My code follows the style guidelines of this project~~
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas that are not self-documenting
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings 
- [x] My changes have no unnecessary logging
- [x] I have added tests that prove my fix is effective or that my feature works, for sufficiently complex features
- [ ] ~~New and existing unit tests pass locally with my changes~~
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Sensitive info like tokens, secrets, and passwords have been removed before submitting

Modified from this article:
Phillip Johnston, “A GitHub Pull Request Template for Your Projects - Embedded Artistry,” Embedded Artistry, Aug. 04, 2017. https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ (accessed Jul. 22, 2021).

**Comment by parseccentric** [comment]
@Barasakar Could you review this when you have a chance?

**Comment by Barasakar** [comment]
@parseccentric  Yeah, I have reviewed this whole thread. Let me know if there is anything else I need to do.

**Comment by parseccentric** [comment]
@Barasakar I made another mistake. As you can see above, I mistakenly committed 508b74d and 18651ff onto this branch before merging. I apologize for this error. I will close this pull request and open a new one with a `temp` branch that reflects the state of the pull request.

**Pull request closed without merging**